File: Kernel Simulator/Misc/Screensaver/Displays/FaderBack.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace Misc.Screensaver.Displays
    Public Module FaderBackSettings

        Private _faderBackDelay As Integer = 10
        Private _faderBackFadeOutDelay As Integer = 3000
        Private _faderBackMaxSteps As Integer = 25
        Private _faderBackMinimumRedColorLevel As Integer = 0
        Private _faderBackMinimumGreenColorLevel As Integer = 0
        Private _faderBackMinimumBlueColorLevel As Integer = 0
        Private _faderBackMaximumRedColorLevel As Integer = 255
        Private _faderBackMaximumGreenColorLevel As Integer = 255
        Private _faderBackMaximumBlueColorLevel As Integer = 255

        ''' <summary>
        ''' [FaderBack] How many milliseconds to wait before making the next write?
        ''' </summary>
        Public Property FaderBackDelay As Integer
            Get
                Return _faderBackDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 10
                _faderBackDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [FaderBack] How many milliseconds to wait before fading the text out?
        ''' </summary>
        Public Property FaderBackFadeOutDelay As Integer
            Get
                Return _faderBackFadeOutDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 3000
                _faderBackFadeOutDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [FaderBack] How many fade steps to do?
        ''' </summary>
        Public Property FaderBackMaxSteps As Integer
            Get
                Return _faderBackMaxSteps
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 25
                _faderBackMaxSteps = value
            End Set
        End Property
        ''' <summary>
        ''' [FaderBack] The minimum red color level (true color)
        ''' </summary>
        Public Property FaderBackMinimumRedColorLevel As Integer
            Get
                Return _faderBackMinimumRedColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _faderBackMinimumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FaderBack] The minimum green color level (true color)
        ''' </summary>
        Public Property FaderBackMinimumGreenColorLevel As Integer
            Get
                Return _faderBackMinimumGreenColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _faderBackMinimumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FaderBack] The minimum blue color level (true color)
        ''' </summary>
        Public Property FaderBackMinimumBlueColorLevel As Integer
            Get
                Return _faderBackMinimumBlueColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _faderBackMinimumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FaderBack] The maximum red color level (true color)
        ''' </summary>
        Public Property FaderBackMaximumRedColorLevel As Integer
            Get
                Return _faderBackMaximumRedColorLevel
            End Get
            Set(value As Integer)
                If value <= _faderBackMinimumRedColorLevel Then value = _faderBackMinimumRedColorLevel
                If value > 255 Then value = 255
                _faderBackMaximumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FaderBack] The maximum green color level (true color)
        ''' </summary>
        Public Property FaderBackMaximumGreenColorLevel As Integer
            Get
                Return _faderBackMaximumGreenColorLevel
            End Get
            Set(value As Integer)
                If value <= _faderBackMinimumGreenColorLevel Then value = _faderBackMinimumGreenColorLevel
                If value > 255 Then value = 255
                _faderBackMaximumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FaderBack] The maximum blue color level (true color)
        ''' </summary>
        Public Property FaderBackMaximumBlueColorLevel As Integer
            Get
                Return _faderBackMaximumBlueColorLevel
            End Get
            Set(value As Integer)
                If value <= _faderBackMinimumBlueColorLevel Then value = _faderBackMinimumBlueColorLevel
                If value > 255 Then value = 255
                _faderBackMaximumBlueColorLevel = value
            End Set
        End Property

    End Module

    Public Class FaderBackDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random
        Private CurrentWindowWidth As Integer
        Private CurrentWindowHeight As Integer
        Private ResizeSyncing As Boolean
        Private FaderBackSettingsInstance As Animations.FaderBack.FaderBackSettings

        Public Overrides Property ScreensaverName As String = "FaderBack" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
            Console.BackgroundColor = ConsoleColor.Black
            ConsoleWrapper.Clear()
            Wdbg(DebugLevel.I, "Console geometry: {0}x{1}", ConsoleWrapper.WindowWidth, ConsoleWrapper.WindowHeight)
            FaderBackSettingsInstance = New Animations.FaderBack.FaderBackSettings With {
                .FaderBackDelay = FaderBackDelay,
                .FaderBackFadeOutDelay = FaderBackFadeOutDelay,
                .FaderBackMaxSteps = FaderBackMaxSteps,
                .FaderBackMinimumRedColorLevel = FaderBackMinimumRedColorLevel,
                .FaderBackMinimumGreenColorLevel = FaderBackMinimumGreenColorLevel,
                .FaderBackMinimumBlueColorLevel = FaderBackMinimumBlueColorLevel,
                .FaderBackMaximumRedColorLevel = FaderBackMaximumRedColorLevel,
                .FaderBackMaximumGreenColorLevel = FaderBackMaximumGreenColorLevel,
                .FaderBackMaximumBlueColorLevel = FaderBackMaximumBlueColorLevel,
                .RandomDriver = RandomDriver
            }
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            Animations.FaderBack.Simulate(FaderBackSettingsInstance)
        End Sub

    End Class
End Namespace

File: Kernel Simulator/Misc/Screensaver/Displays/FaderBack.cs
using System;
using System.Collections.Generic;
using KS.Misc.Writers.DebugWriters;
using Terminaux.Base;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

namespace KS.Misc.Screensaver.Displays
{
	public static class FaderBackSettings
	{

		private static int _faderBackDelay = 10;
		private static int _faderBackFadeOutDelay = 3000;
		private static int _faderBackMaxSteps = 25;
		private static int _faderBackMinimumRedColorLevel = 0;
		private static int _faderBackMinimumGreenColorLevel = 0;
		private static int _faderBackMinimumBlueColorLevel = 0;
		private static int _faderBackMaximumRedColorLevel = 255;
		private static int _faderBackMaximumGreenColorLevel = 255;
		private static int _faderBackMaximumBlueColorLevel = 255;

		/// <summary>
        /// [FaderBack] How many milliseconds to wait before making the next write?
        /// </summary>
		public static int FaderBackDelay
		{
			get
			{
				return _faderBackDelay;
			}
			set
			{
				if (value <= 0)
					value = 10;
				_faderBackDelay = value;
			}
		}
		/// <summary>
        /// [FaderBack] How many milliseconds to wait before fading the text out?
        /// </summary>
		public static int FaderBackFadeOutDelay
		{
			get
			{
				return _faderBackFadeOutDelay;
			}
			set
			{
				if (value <= 0)
					value = 3000;
				_faderBackFadeOutDelay = value;
			}
		}
		/// <summary>
        /// [FaderBack] How many fade steps to do?
        /// </summary>
		public static int FaderBackMaxSteps
		{
			get
			{
				return _faderBackMaxSteps;
			}
			set
			{
				if (value <= 0)
					value = 25;
				_faderBackMaxSteps = value;
			}
		}
		/// <summary>
        /// [FaderBack] The minimum red color level (true color)
        /// </summary>
		public static int FaderBackMinimumRedColorLevel
		{
			get
			{
				return _faderBackMinimumRedColorLevel;
			}
			set
			{
				if (value <= 0)
					value = 0;
				if (value > 255)
					value = 255;
				_faderBackMinimumRedColorLevel = value;
			}
		}
		/// <summary>
        /// [FaderBack] The minimum green color level (true color)
        /// </summary>
		public static int FaderBackMinimumGreenColorLevel
		{
			get
			{
				return _faderBackMinimumGreenColorLevel;
			}
			set
			{
				if (value <= 0)
					value = 0;
				if (value > 255)
					value = 255;
				_faderBackMinimumGreenColorLevel = value;
			}
		}
		/// <summary>
        /// [FaderBack] The minimum blue color level (true color)
        /// </summary>
		public static int FaderBackMinimumBlueColorLevel
		{
			get
			{
				return _faderBackMinimumBlueColorLevel;
			}
			set
			{
				if (value <= 0)
					value = 0;
				if (value > 255)
					value = 255;
				_faderBackMinimumBlueColorLevel = value;
			}
		}
		/// <summary>
        /// [FaderBack] The maximum red color level (true color)
        /// </summary>
		public static int FaderBackMaximumRedColorLevel
		{
			get
			{
				return _faderBackMaximumRedColorLevel;
			}
			set
			{
				if (value <= _faderBackMinimumRedColorLevel)
					value = _faderBackMinimumRedColorLevel;
				if (value > 255)
					value = 255;
				_faderBackMaximumRedColorLevel = value;
			}
		}
		/// <summary>
        /// [FaderBack] The maximum green color level (true color)
        /// </summary>
		public static int FaderBackMaximumGreenColorLevel
		{
			get
			{
				return _faderBackMaximumGreenColorLevel;
			}
			set
			{
				if (value <= _faderBackMinimumGreenColorLevel)
					value = _faderBackMinimumGreenColorLevel;
				if (value > 255)
					value = 255;
				_faderBackMaximumGreenColorLevel = value;
			}
		}
		/// <summary>
        /// [FaderBack] The maximum blue color level (true color)
        /// </summary>
		public static int FaderBackMaximumBlueColorLevel
		{
			get
			{
				return _faderBackMaximumBlueColorLevel;
			}
			set
			{
				if (value <= _faderBackMinimumBlueColorLevel)
					value = _faderBackMinimumBlueColorLevel;
				if (value > 255)
					value = 255;
				_faderBackMaximumBlueColorLevel = value;
			}
		}

	}

	public class FaderBackDisplay : BaseScreensaver, IScreensaver
	{

		private Random RandomDriver;
		private int CurrentWindowWidth;
		private int CurrentWindowHeight;
		private bool ResizeSyncing;
		private Animations.FaderBack.FaderBackSettings FaderBackSettingsInstance;

		public override string ScreensaverName { get; set; } = "FaderBack";

		public override Dictionary<string, object> ScreensaverSettings { get; set; }

		public override void ScreensaverPreparation()
		{
			// Variable preparations
			RandomDriver = new Random();
			CurrentWindowWidth = ConsoleWrapper.WindowWidth;
			CurrentWindowHeight = ConsoleWrapper.WindowHeight;
			Console.BackgroundColor = ConsoleColor.Black;
			ConsoleWrapper.Clear();
			DebugWriter.Wdbg(DebugLevel.I, "Console geometry: {0}x{1}", ConsoleWrapper.WindowWidth, ConsoleWrapper.WindowHeight);
			FaderBackSettingsInstance = new Animations.FaderBack.FaderBackSettings()
			{
				FaderBackDelay = FaderBackSettings.FaderBackDelay,
				FaderBackFadeOutDelay = FaderBackSettings.FaderBackFadeOutDelay,
				FaderBackMaxSteps = FaderBackSettings.FaderBackMaxSteps,
				FaderBackMinimumRedColorLevel = FaderBackSettings.FaderBackMinimumRedColorLevel,
				FaderBackMinimumGreenColorLevel = FaderBackSettings.FaderBackMinimumGreenColorLevel,
				FaderBackMinimumBlueColorLevel = FaderBackSettings.FaderBackMinimumBlueColorLevel,
				FaderBackMaximumRedColorLevel = FaderBackSettings.FaderBackMaximumRedColorLevel,
				FaderBackMaximumGreenColorLevel = FaderBackSettings.FaderBackMaximumGreenColorLevel,
				FaderBackMaximumBlueColorLevel = FaderBackSettings.FaderBackMaximumBlueColorLevel,
				RandomDriver = RandomDriver
			};
		}

		public override void ScreensaverLogic()
		{
			Animations.FaderBack.FaderBack.Simulate(FaderBackSettingsInstance);
		}

	}
}